buildSrc/src/main/kotlin/conventions/publishing.gradle.kts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package conventions
+
+// Provides the publishing configuration for the build
+//
+// Note: Further configuration can be achieved using the `project.configureMavenPublication` and
+// `project.configureJarManifest` helpers.
+// See: `ProjectExtensions.kt` for more information
+plugins {
+    id("conventions.git-version")
+    id("maven-publish")
+    id("signing")
+}
+
+val signingKey: Provider<String> = providers.gradleProperty("signingKey")
+val signingPassword: Provider<String> = providers.gradleProperty("signingPassword")
+val nexusUsername: Provider<String> = providers.gradleProperty("nexusUsername")
+val nexusPassword: Provider<String> = providers.gradleProperty("nexusPassword")
+@Suppress("UNCHECKED_CAST") val gitVersion: Provider<String> = project.findProperty("gitVersion") as Provider<String>
+
+tasks.withType<AbstractPublishToMaven>().configureEach {
+    // Gradle warns about some signing tasks using publishing task outputs without explicit
+    // dependencies. Here's a quick fix.
+    dependsOn(tasks.withType<Sign>())
+    mustRunAfter(tasks.withType<Sign>())
+
+    doLast {
+        logger.lifecycle("[task: ${name}] ${publication.groupId}:${publication.artifactId}:${publication.version}")
+    }
+}
+
+val localBuildRepo: Provider<Directory> = rootProject.layout.buildDirectory.dir("repo")
+
+val sonatypeRepositoryReleaseUrl: Provider<String> = provider {
+    if (version.toString().endsWith("SNAPSHOT")) {
+        "https://oss.sonatype.org/content/repositories/snapshots/"
+    } else {
+        "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
+    }
+}
+
+publishing {
+    repositories {
+        maven {
+            url = uri(sonatypeRepositoryReleaseUrl)
+            if (nexusUsername.isPresent && nexusPassword.isPresent) {
+                credentials {
+                    username = nexusUsername.get()
+                    password = nexusPassword.get()
+                }
+            }
+        }
+
+        // publish to local dir, for artifact tracking and testing
+        // `./gradlew publishMavenPublicationToLocalBuildRepository`
+        maven {
+            url = uri(localBuildRepo.get())
+            name = "LocalBuild"
+        }
+    }
+
+    publications.create<MavenPublication>("maven") {
+        components.findByName("scala")?.let { from(it) }
```

Comment:
I'm not clear what "scala" and "java" are referring to in this context.  I would also expect to see "kotlin", so I'm even less sure what this is about.

Reply:
Turns out scala is not needed / doesn't exist - I just ported it but testing shows it can be removed.

Java components are as essentially that the Jar that the java/scala/kotlin plugins create.